source/common/http/alternate_protocols.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "envoy/common/time.h"
+
+namespace Envoy {
+namespace Http {
+
+// Tracks alternate protocols that can be used to make an HTTP connection to an origin server.
+class AlternateProtocols {
+public:
+  // Represents an HTTP origin to be connected too.
+  struct Origin {
+  public:
+    Origin(std::string scheme, std::string hostname, int port);
```

Comment:
nit: prefer passing `absl::string_view` instead of `std::string` to functions to avoid copies

Reply:
Oh, yes of course! My C++ is rusty! :( Done.